### ISSUE TITLE
FileNotFoundException when loading EF6 Migrations

### ISSUE THREAD

**Issue by mklaber** [aspnet/Home#1212]
This is an issue that is similar to #998 and was alluded to in a great blog post called [Entity Framework 6 With ASP.NET 5](http://bleedingnedge.com/2015/11/01/entity-framework-6-with-asp-net-5/) by @pgrudzien12. I've [forked his repository](https://github.com/mklaber/Examples.EntityFramework) and updated it to use `rc1-final`. I've also included the specific exception that happens upon running the web app.

In short, manual EF migrations (e.g., ones created using `Add-Migration`) cause the following exception: `FileNotFoundException: Could not find file '{Ef Project Namespace}.resources'.` I've replicated this by just starting the web app. I've also tried (in another project) manually calling `Update()` from an instance of `DbMigrator`.

I've also attempted, per #998 to instantiate the [`ResourceManager`](https://github.com/mklaber/Examples.EntityFramework/blob/master/DataAccess/Migrations/201510241112573_InitialDatabase.Designer.cs#L12) more explicitly: `private readonly ResourceManager Resources = new ResourceManager(typeof (InitialDatabase).FullName, System.Reflection.Assembly.Load(typeof (InitialDatabase).Assembly.GetName()));`with no success.

Here's the full stacktrace:

```
System.IO.FileNotFoundException: Could not find file 'WebApplication.DataAccess.resources'.
   at System.Reflection.RuntimeAssembly.InternalGetSatelliteAssembly(String name, CultureInfo culture, Version version, Boolean throwOnFileNotFound, StackCrawlMark& stackMark)
   at System.Resources.ManifestBasedResourceGroveler.GetSatelliteAssembly(CultureInfo lookForCulture, StackCrawlMark& stackMark)
   at System.Resources.ManifestBasedResourceGroveler.GrovelForResourceSet(CultureInfo culture, Dictionary`2 localResourceSets, Boolean tryParents, Boolean createIfNotExists, StackCrawlMark& stackMark)
   at System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo requestedCulture, Boolean createIfNotExists, Boolean tryParents, StackCrawlMark& stackMark)
   at System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo culture, Boolean createIfNotExists, Boolean tryParents)
   at System.Resources.ResourceManager.GetString(String name, CultureInfo culture)
   at System.Resources.ResourceManager.GetString(String name)
   at System.Data.Entity.Migrations.DbMigrator.GetDefaultSchema(DbMigration migration)
   at System.Data.Entity.Migrations.DbMigrator.<GetHistorySchemas>b__6(<>f__AnonymousType10`2 <>h__TransparentIdentifier4)
   at System.Linq.Enumerable.<>c__DisplayClass7_0`3.<CombineSelectors>b__0(TSource x)
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<DistinctIterator>d__63`1.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.<ReverseIterator>d__74`1.MoveNext()
   at System.Data.Entity.Migrations.History.HistoryRepository.QueryExists(String contextKey)
   at System.Data.Entity.Migrations.History.HistoryRepository.Exists(String contextKey)
   at System.Data.Entity.Migrations.History.HistoryRepository.<GetUpgradeOperations>d__16.MoveNext()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source)
   at System.Data.Entity.Migrations.DbMigrator.UpdateInternal(String targetMigration)
   at System.Data.Entity.Migrations.DbMigrator.<>c__DisplayClassc.<Update>b__b()
   at System.Data.Entity.Migrations.DbMigrator.EnsureDatabaseExists(Action mustSucceedToKeepDatabase)
   at System.Data.Entity.Migrations.Infrastructure.MigratorBase.EnsureDatabaseExists(Action mustSucceedToKeepDatabase)
   at System.Data.Entity.Migrations.DbMigrator.Update(String targetMigration)
   at System.Data.Entity.Migrations.Infrastructure.MigratorBase.Update()
   at System.Data.Entity.Internal.DatabaseCreator.CreateDatabase(InternalContext internalContext, Func`3 createMigrator, ObjectContext objectContext)
   at System.Data.Entity.Internal.InternalContext.CreateDatabase(ObjectContext objectContext, DatabaseExistenceState existenceState)
   at System.Data.Entity.Database.Create(DatabaseExistenceState existenceState)
   at System.Data.Entity.CreateDatabaseIfNotExists`1.InitializeDatabase(TContext context)
   at System.Data.Entity.Internal.InternalContext.<>c__DisplayClassf`1.<CreateInitializationAction>b__e()
   at System.Data.Entity.Internal.InternalContext.PerformInitializationAction(Action action)
   at System.Data.Entity.Internal.InternalContext.PerformDatabaseInitialization()
   at System.Data.Entity.Internal.LazyInternalContext.<InitializeDatabase>b__4(InternalContext c)
   at System.Data.Entity.Internal.RetryAction`1.PerformAction(TInput input)
   at System.Data.Entity.Internal.LazyInternalContext.InitializeDatabaseAction(Action`1 action)
   at System.Data.Entity.Internal.LazyInternalContext.InitializeDatabase()
   at System.Data.Entity.Internal.InternalContext.Initialize()
   at System.Data.Entity.Internal.InternalContext.GetEntitySetAndBaseTypeForType(Type entityType)
   at System.Data.Entity.Internal.Linq.InternalSet`1.Initialize()
   at System.Data.Entity.Internal.Linq.InternalSet`1.GetEnumerator()
   at System.Data.Entity.Infrastructure.DbQuery`1.System.Collections.Generic.IEnumerable<TResult>.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at WebApplication.Mvc6.Controllers.HomeController.Index() in C:\Code\Examples.EntityFramework\WebApplication.Mvc6\Controllers\HomeController.cs:line 21
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNet.Mvc.Controllers.ControllerActionExecutor.ExecuteAsync(MethodInfo actionMethodInfo, Object instance, Object[] orderedActionArguments)
   at Microsoft.AspNet.Mvc.Controllers.ControllerActionInvoker.<InvokeActionAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.AspNet.Mvc.Controllers.FilterActionInvoker.<InvokeActionFilterAsync>d__53.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.AspNet.Mvc.Controllers.FilterActionInvoker.<InvokeAsync>d__44.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.AspNet.Mvc.Infrastructure.MvcRouteHandler.<RouteAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.AspNet.Routing.Template.TemplateRoute.<RouteAsync>d__27.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.AspNet.Routing.RouteCollection.<RouteAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.AspNet.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.AspNet.IISPlatformHandler.IISPlatformHandlerMiddleware.<Invoke>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Microsoft.AspNet.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__7.MoveNext()
```

My only workaround is to start an MVC 5 app (which I fortunately still have in my solution) just to run the migration.

My only theory about what this could be related to is perhaps a limitation in the [`project.json`](https://github.com/mklaber/Examples.EntityFramework/blob/master/wrap/WebApplication.DataAccess/project.json) wrap?

My apologies, in advance, if I've not posted to this to the most specific repository. If there's a better home for this issue, please let me know and I'll re-post.

**Comment by davidfowl** [comment]
What's the name of the resource file in the application?

**Comment by mklaber** [comment]
The exception is happening with this call to `Resources.GetString`: [201510241112573_InitialDatabase.Designer.cs](https://github.com/mklaber/Examples.EntityFramework/blob/master/DataAccess/Migrations/201510241112573_InitialDatabase.Designer.cs#L24-L27)

That's in the designer file for [201510241112573_InitialDatabase.resx](https://github.com/mklaber/Examples.EntityFramework/blob/master/DataAccess/Migrations/201510241112573_InitialDatabase.resx)

Does that answer your question?

**Comment by davidfowl** [comment]
I cloned the repository, built the project and opened ILSpy. It looks like you're just using the wrong resource name:

![image](https://cloud.githubusercontent.com/assets/95136/12119125/a95c9f5a-b37e-11e5-8be1-a2f12d92f79b.png)

**Comment by mklaber** [comment]
All of the code in [201510241112573_InitialDatabase.Designer.cs](https://github.com/mklaber/Examples.EntityFramework/blob/master/DataAccess/Migrations/201510241112573_InitialDatabase.Designer.cs) is auto generated by EF migrations. While I suppose I could manually edit this file to use a different resource name (not sure where I'd set that—perhaps manually `typeof(InitialDatabase).FullName + ".resources", typeof(InitialDatabase).Assembly`?), this is going to be a problem for anyone loading EF generated migrations.


In other words, if I'm using the wrong resource name, so is anyone else who is loading an EF6 DbContext with manual migrations.

**Comment by mklaber** [comment]
I updated the auto-gen code in the Designer file to this:

`private readonly ResourceManager Resources = new ResourceManager(typeof(InitialDatabase).FullName + ".resources", typeof(InitialDatabase).Assembly);`

And am still getting the same error (even though `typeof(InitialDatabase).FullName + ".resources"` does resolve to the name in your screenshot).

**Comment by davidfowl** [comment]
Ah I thought it was `nameof` not `typeof`. That's pretty strange. Will take a deeper look

**Comment by mklaber** [comment]
Thanks @davidfowl

**Comment by davidfowl** [comment]
Seems like we have an issue with the way we resolve resources in DNX. This will be fixed for RC2 when we move to the CLI so it'll fix itself :smile:. As for a workaround, I haven't found a good one as yet. I do know that passing invariant culture works though...

**Comment by mklaber** [comment]
@davidfowl how'd you modify the call to `GetString` to get it to work? I just tried this:

```csharp
        string IMigrationMetadata.Target
        {
            get { return Resources.GetString("Target", System.Globalization.CultureInfo.InvariantCulture); }
        }
```

And am getting the same darn exception.

If I can get that to work, seems like a sufficient workaround for now.

Is the ETA (or is it ETR?) of the rc2 release still [Feb 2016](https://github.com/aspnet/Home/wiki/Roadmap#schedule)?

**Comment by davidfowl** [comment]
Hmmm, that seemed to work for me when I tested the same code in the ASP.NET 5 app.

> Is the ETA (or is it ETR?) of the rc2 release still Feb 2016?

Yep. You can start trying to use the new toolchain now but it's a work in progress. e.g. https://github.com/davidfowl/dotnetcli-aspnet5

**Comment by mklaber** [comment]
Is it possible your test didn't actually hit that line? I was having a problem troubleshooting because in some cases (which cases specifically was not clear) it skipped the migration step and thus appeared to load. However, the changes that came as part of an EF migration were not made. I worked around that by making sure to change the name of the Initial Catalog each time I ran a test.

**Comment by davidfowl** [comment]
Here's what I did in Startup:

```C#
public void ConfigureServices(IServiceCollection services)
        {

            // services.AddScoped<IPostRepository, SqlPostRepository>();
            // services.AddScoped((_) => ApplicationDbContext.Create(Configuration["Data:DefaultConnection:ConnectionString"]));
            // var rm = new ResourceManager(typeof(InitialDatabase));
            var rm = new ResourceManager(typeof(InitialDatabase));
            var target = rm.GetString("Target", CultureInfo.InvariantCulture);
            // Add Entity Framework services to the services container.
            //services.AddEntityFramework()
            //    .AddSqlServer()
            //    .AddDbContext<ApplicationDbContext>(options =>
            //        options.UseSqlServer(Configuration["Data:DefaultConnection:ConnectionString"]));



            // Add MVC services to the services container.
            services.AddMvc();
        }
```

**Comment by mklaber** [comment]
Ah, yea, that works for me too. Just doesn't work from the WebApplication.DataAccess library. :-(

**Comment by mklaber** [comment]
Super duper duper duper hack but here's what I did to get some integration tests working:

```csharp
        private void Migrate<TDbContextAssembly>(string connectionString)
        {
            var efAssemblyLocation = Path.GetDirectoryName(typeof (DbContext).Assembly.Location);

            var migrateToolToCopy = Path.Combine(efAssemblyLocation, @"..\..\tools", "migrate.exe");
            var migrateTool = Path.Combine(efAssemblyLocation, "migrate.exe");

            if (!File.Exists(migrateTool))
                File.Copy(migrateToolToCopy, migrateTool);

            var contextDllLocation = typeof (TDbContextAssembly).Assembly.Location.Replace(@"\obj\", @"\bin\");
            var contextDllPath = Path.GetDirectoryName(contextDllLocation);
            var contextDllName = Path.GetFileName(contextDllLocation);
            int timeout = (int) TimeSpan.FromMinutes(1).TotalMilliseconds;

            // process handling from: http://stackoverflow.com/a/7608823/612166
            using (var process = new Process()
            {
                EnableRaisingEvents = true,
                StartInfo = new ProcessStartInfo(migrateTool,
                    $"{contextDllName} " +
                    $"/StartUpDirectory:{contextDllPath} " +
                    $"/connectionString:\"{connectionString}\" " +
                    $"/connectionProviderName:System.Data.SqlClient " +
                    $"/Verbose")
                {
                    UseShellExecute = false,
                    RedirectStandardError = true,
                    RedirectStandardOutput = true,
                    WorkingDirectory = Path.GetDirectoryName(migrateTool),
                }
            })
            {
                StringBuilder output = new StringBuilder();
                StringBuilder error = new StringBuilder();

                using (AutoResetEvent outputWaitHandle = new AutoResetEvent(false))
                using (AutoResetEvent errorWaitHandle = new AutoResetEvent(false))
                {
                    process.OutputDataReceived += (sender, e) =>
                    {
                        if (e.Data == null)
                        {
                            outputWaitHandle.Set();
                        }
                        else
                        {
                            output.AppendLine(e.Data);
                        }
                    };
                    process.ErrorDataReceived += (sender, e) =>
                    {
                        if (e.Data == null)
                        {
                            errorWaitHandle.Set();
                        }
                        else
                        {
                            error.AppendLine(e.Data);
                        }
                    };

                    process.Start();

                    process.BeginOutputReadLine();
                    process.BeginErrorReadLine();

                    if (process.WaitForExit(timeout) &&
                        outputWaitHandle.WaitOne(timeout) &&
                        errorWaitHandle.WaitOne(timeout))
                    {
                        // Process completed. Check process.ExitCode here.
                        process.ExitCode.Should()
                            .Be(0,
                                $"Exit code should have been 0; but was: {error.ToString()} with standard output: {output.ToString()}");
                    }
                    else
                    {
                        // Timed out.
                        false.Should().BeTrue($"Migration didn't finish within allotted timeout: {output}");
                    }
                }
            }
        }
```

Usage: `Migrate<MyDbContext>("Data Source=...");`

**Comment by davidfowl** [comment]
Hopwfully this is fixed with vs2017 and the shift back to msbuild
